keps/sig-node/4033-group-driver-detection-over-cri/README.md

```diff
@@ -0,0 +1,890 @@
+which cgroup driver to use. This removes the need for specifying cgroup driver
+in the kubelet configuration and eliminates the possibility of misaligned
+cgroup driver configuration between the kubelet and the runtime.
+
+## Motivation
+
+The responsibility of managing the Linux cgroups is currently split between the
+kubelet and the container runtime. Kubelet takes care of the pod (sandbox)
+level cgroups whereas the runtime is responsible for per-container cgroups.
+There currently are two different low-level management interfaces for cgroups:
+maninpulating the cgroupfs directly or using the systemd system management
+daemon to manage them. Currently, both the kubelet and the container runtime
+has a configuration setting for selecting the cgroup driver (cgroupfs or
+systemd). These settings must be in sync, both kubelet and the runtime
+configured to use the same driver as the two drivers are incompatible because
+of a different kind of cgroups hierarchy used in them. Having kubelet and the
+container runtime to use non-matching cgroup drivers can cause hard to
+understand failures in container creation or inconsistent resource allocation
+on the node. This – two independent configuration settings for the same thing –
+is a common cause for user errors. Instead of having a split brain situation,
+there should be a single source of truth for the cgroup driver.
+
+### Goals
+
+- make kubelet automatically use the same cgroup driver as the container
+  runtime, making it unnecessary to specify cgroup driver in kubelet
+  configuration
+- maintain backward compatibility to run kubelet on top of an older container
+  runtime that doesn’t support this new feature
+
+### Non-Goals
+
+## Proposal
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a cluster administrator, I would like to simplify my node configuration by
+configuring fields just once, and not needing to synchronize between multiple
+processes.
+
+#### Story 2
+
+As a novice Kubernetes user, I would like to easily be able to start pods with
+all runtimes, even if they have differing opinions on defaults.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+Field adoption could be considered a risk, though the CRI implementations work
+closely with SIG Node and the feature will move along with CRI implementation
+adoption.
+
+## Design Details
+
+### CRI API
+
+Extend the CRI protocol to inform the kubelet which cgroup driver should be
+used.
+
+```diff
+ message RuntimeStatus {
+     // List of current observed runtime conditions.
+     repeated RuntimeCondition conditions = 1;
++    // Configuration settings of the runtime
++    RuntimeConfiguration configuration = 2;
```

Comment:
I think we should split this into a separate CRI call.

Reply:
@mrunalp I think in this case in this case we should add a streaming interface for the runtime to be able to notify about dynamic changes (not for the cgroup driver but for other potentially more dynamic stuff in the future). Adding yet another endpoint that get's periodically polled doesn't sound reasonable to me. WDYT? @haircommander ?

Reply:
yeah I think something worth noting is we're not preparred in this KEP to react to the runtime changing its mind. I don't think the kubelet should update the cgroup driver if the cri changes without a kubelet restart. from that perspective, separating the calls may be appropriate

Reply:
> I don't think the kubelet should update the cgroup driver if the cri changes

I agree. But I was thinking this from the CRI API pov. So we would add the simple "unary" rpc for now and then later with the dynamic stuff add a parallel streaming rpc (for basically the same data, I think). Is this the way to go?

Reply:
yeah I find myself leaning towards having a separate API call for the streaming stuff, part to reduce the info going over the wire, part to make clear how the kubelet will react to the data

Reply:
OK, changed the proposal to add a new `RuntimeConfig` rpc for querying the runtime configuration. @mrunalp @haircommander WDYT?